core/services/ocr2/plugins/ocr2keeper/evm21/upkeepstate/orm.go

```diff
@@ -0,0 +1,79 @@
+package upkeepstate
+
+import (
+	"database/sql"
+	"errors"
+	"math/big"
+	"time"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pg"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+	"github.com/smartcontractkit/sqlx"
+)
+
+type ORM struct {
+	chainID *big.Int
+	q       pg.Q
+}
+
+// NewORM creates an ORM scoped to chainID.
+func NewORM(chainID *big.Int, db *sqlx.DB, lggr logger.Logger, cfg pg.QConfig) *ORM {
+	return &ORM{
+		chainID: chainID,
+		q:       pg.NewQ(db, lggr.Named("ORM"), cfg),
+	}
+}
+
+// InsertUpkeepState is idempotent and sets upkeep state values in db
+func (o *ORM) InsertUpkeepState(state upkeepStateRecord, qopts ...pg.QOpt) error {
+	q := o.q.WithOpts(qopts...)
+
+	query := `INSERT INTO evm_upkeep_state (evm_chain_id, work_id, completion_state, block_number, added_at)
+	  VALUES ($1::NUMERIC, $2, $3, $4::NUMERIC, $5)
+	    ON CONFLICT (evm_chain_id, work_id)
+	    DO UPDATE SET completion_state=$3, block_number=$4`
+
+	return q.ExecQ(query, utils.NewBig(o.chainID), state.WorkID, state.CompletionState, state.BlockNumber, state.AddedAt)
+}
+
+// SelectStatesByWorkIDs searches the data store for stored states for the
+// provided work ids and configured chain id
+func (o *ORM) SelectStatesByWorkIDs(workIDs []string, qopts ...pg.QOpt) (states []upkeepStateRecord, err error) {
```

Comment:
create your on struct (or use check result) instead of using the cache record which don't has all the needed fields